### DISCUSSION TITLE
Porting to latest LibreMetaverse, issues coming from a very old version.

### DISCUSSION THREAD

**Discussion by mdickson** [cinderblocks/libremetaverse#69] [General]
Porting the Halcyon grid code (soon to be Tranquility grid code) to the latest Libremetaverse.  I worked past the Image handling issues with layer info and there are 2 left.  BlockingQueue is missing and It looks like OutgoingPacket has moved to NetworkManager and changed.  IDK if you have any idea if its still used the same, the big change for my part looks to be a missing reference to the client.  it does include a scene reference and both were just objects so I may be able to reuse that.  

This was previously built with v1.0.2.26 so lots of room for changes.  Is there any existing opensim that compiles with the current library that I could use as a reference?  

Any advice is appreciated.  Thye Blocking Queue reference.. Well I can probably find something to replace that.   But again any suggestions are appreciated.

**Comment by cinderblocks** [comment]
`BlockingQueue`, `CircularQueue`, and `LocklessQueue` were removed. Use C# containers instead. Almost all uses in the library itself were replaced with `System.Collections.ConcurrentQueue<T>` with `SemaphoreSlim` used for checking size and waiting.

`OutgoingPacket` did change. `Client` was changed to `Simulator`. It's possible this was being used in OpenSim, but usage was identical within `Simulator` just without the need to cast to and from `Object`.

**Comment by mdickson** [comment]
Well it hasnt been used in OpenSim (until now :).  Simulator is correct from a client POV but in this case its an outbound packet from the simulator to a client,  OTOH I totally agree there has to be a better way than a generic object and casting.  I can sort this out for now and then figure out how to push it down to the library later.  The Comms stack in LibreMetaverse probably really is more client to server focused so maybe the outbound code in the simulator shouldn't be using it.

**Comment by cinderblocks** [comment]
I'm not opposed to the idea of segmenting out the shared functionality and definitions and creating a more server-focused library either. Especially with .NET Core support on not-windows being so much better these days.
